### DISCUSSION TITLE
provider_refresh_token is always null for github oauth provider

### DISCUSSION THREAD

**Discussion by Warchant** [supabase/supabase#18399] [Questions] [answered]
I use Auth component like this:
```tsx
<Auth
            supabaseClient={supabase}
            providers={['github']}
            appearance={{ theme: ThemeSupa }}
            theme="dark"
            showLinks={false}
            redirectTo={`http://localhost:3000/api/auth/callback`}
            onlyThirdPartyProviders
        />
```

`/api/auth/callback/route.ts`:
```typescript
export async function GET(req: NextRequest) {
  const cookieStore = cookies();
  const supabase = createRouteHandlerClient({ cookies: () => cookieStore });
  const { searchParams } = new URL(req.url);
  const code = searchParams.get("code");
  if (code) {
    await supabase.auth.exchangeCodeForSession(code);
  }

  return NextResponse.redirect(new URL("/", req.url));
}
```

When I click "Sign In with Github" on a frontend, my server receives a session with a valid `provider_token` (from github), but `provider_refresh_token` is always `null`. 
```json
  "provider_token": "<actual valid token>",
  "provider_refresh_token": null,
  "user": {
    ...
    "app_metadata": {
      "provider": "github",
      "providers": [
        "github"
      ]
    },
```

The question is - is it a bug? I expect github to return a refresh_token so that I can refresh provider_token without explicit user signout/signin.

**Comment by GaryAustin1** [comment]
Do you have the refresh option set in Github?  https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/refreshing-user-access-tokens

**Reply by Warchant** [reply]
@GaryAustin1 I am using `OAuth App` (not GitHub App). And "Optional Features" page does not have anything:
![image](https://github.com/supabase/supabase/assets/1867551/b1421ff3-4981-48b0-812f-d22959cab4cf)

**Reply by GaryAustin1** [reply]
Are you sure there is supposed to be a refresh token for Github OAuth?
https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps
![image](https://github.com/supabase/supabase/assets/54564956/f3ff34ea-f060-4a53-a347-6c2ea20baaeb)

See several comments on the web that it does not use refresh tokens, but I've not used it so just going on what I see.

**Reply by Warchant** [reply]
@GaryAustin1 I am not sure if that refresh token is supposed to be there. 

I raised this issue mainly because `provider_token` gets expired pretty quickly. After that, session stores `provider_token: null` and `provider_refresh_token: null`, and I am not sure how to get `provider_token` again... 
It seems that I need to reauthenticate a user.

**Reply by GaryAustin1** [reply, marked as answer]
Supabase does not keep the provider_token in the session.  It is there only after OAuth call and goes away when the session refreshes.  You are responsible for storing it if you need to use it.

**Reply by GaryAustin1** [reply]
https://supabase.com/docs/guides/auth/social-login#provider-tokens

**Reply by dannypurcell** [reply]
The instructions provided at that link: 
```
Supabase Auth does not manage refreshing the provider token for the user. Your application will need to use the provider refresh token to obtain a new provider token.
```
Are not possible to follow because Supabase must be registered with the OAuth provider as a "Web" type app and cross-origin token refresh is typically only allowed for "SPA" type apps which use a `clientSecret` to verify they are allowed to request tokens.

We need a better answer than this for signing in with a provider while also needing to make requests to the provider's own APIs.

Right now the only answer is to have the user sign in twice. Which is not acceptable from a UX standpoint.

**Reply by dannypurcell** [reply]
This issue is also what this question https://github.com/orgs/supabase/discussions/20035 is actually having problems with.

It is also what many of the people arriving at this discussion are having an issue with:
https://github.com/orgs/supabase/discussions/2961#discussioncomment-4600847

In my own case Microsoft/Azure returns 
```
 AADSTS9002326: Cross-origin token redemption is permitted only for the 'Single-Page Application' client-type. Request origin: 'http://localhost:8080'.
```
When attempting a token refresh using the `providerRefreshToken` from the Supabase auth Session.

**Reply by GaryAustin1** [reply]
@dannypurcell I doubt anyone from Supabase is going to see your posts on older discussions.
You might want to generate an issue in supabase/gotrue if there is not one there yet.

**Reply by dannypurcell** [reply]
Thank for the heads up. I've opened a new discussion as well, https://github.com/orgs/supabase/discussions/20795
Just making notes on the old ones for others who may come across them in searching for help with this issue as I did.
